Public/Src/Engine/Dll/Distribution/RemoteWorker.cs

```diff
@@ -381,58 +381,112 @@ public override async void Start()
-            if (m_sendThread.IsAlive)
+            while (true)
             {
-                m_sendThread.Join();
+                WorkerNodeStatus status = Status;
+                if (status == WorkerNodeStatus.Stopping || status == WorkerNodeStatus.Stopped)
+                {
+                    // We already initiated the stop for the worker.
+                    return;
+                }
+
+                if (ChangeStatus(status, WorkerNodeStatus.Stopping))
+                {
+                    break;
+                }
             }
 
-            bool initiatedStop = false;
+            await DisconnectAsync(buildFailure);
+        }
+
+        /// <inheritdoc />
+        public override async Task EarlyReleaseAsync()
+        {
+            // Unblock scheduler
+            await Task.Yield();
+
             while (true)
             {
                 WorkerNodeStatus status = Status;
                 if (status == WorkerNodeStatus.Stopping || status == WorkerNodeStatus.Stopped)
                 {
-                    break;
+                    // Already stopped, no need to continue.
+                    return;
                 }
 
                 if (ChangeStatus(status, WorkerNodeStatus.Stopping))
                 {
-                    initiatedStop = true;
                     break;
                 }
             }
 
-            if (initiatedStop)
-            {
-                CancellationTokenSource exitCancellation = new CancellationTokenSource();
+            var drainStopwatch = new StopwatchVar();
 
-                // Only wait a short amount of time for exit (15 seconds) if worker is not successfully attached.
-                if (m_attachCompletion.Task.Status != TaskStatus.RanToCompletion || !await m_attachCompletion.Task)
+            using (drainStopwatch.Start())
+            {
+                if (AcquiredSlots != 0)
                 {
-                    exitCancellation.CancelAfter(TimeSpan.FromSeconds(15));
+                    await DrainCompletion.Task;
                 }
+            }
 
-                var buildEndData = new BuildEndData()
-                {
-                    Failure = buildFailure ?? m_exitFailure
-                };
+            Contract.Assert(m_pipCompletionTasks.IsEmpty, "There cannot be pending completion tasks when AcquiredSlots is 0");
 
-                await m_workerClient.ExitAsync(buildEndData, exitCancellation.Token);
+            var disconnectStopwatch = new StopwatchVar();
 
-                m_executionBlobQueue.CompleteAdding();
+            using (disconnectStopwatch.Start())
+            {
+                await DisconnectAsync();
+            }
+
+            Scheduler.Tracing.Logger.Log.WorkerReleasedEarly(
+                m_appLoggingContext,
+                Name,
+                (long)drainStopwatch.TotalElapsed.TotalMilliseconds,
+                (long)disconnectStopwatch.TotalElapsed.TotalMilliseconds);

```

Comment:
Can you add counters for these so they go into the stats file and telemetry? I'm thinking the total drain duration, disconnect duration, and number of workers released early. Please don't just add send the WorkerReleasedEarly event to telemetry too since that will be complicated to join against due to multiple events per session and we're low on our unique telemetry event quota.

Reply:
Add the followings:

RemoteWorker_EarlyReleaseSavingDurationMs
RemoteWorker_EarlyReleaseDrainDurationMs

---
In reply to: [289496810](https://github.com/microsoft/BuildXL/pull/392#discussion_r289496810) [](ancestors = 289496810)